matrix_client/crypto/olm_device.py

```diff
@@ -0,0 +1,94 @@
+import logging
+
+import olm
+from canonicaljson import encode_canonical_json
+
+from matrix_client.checks import check_user_id
+
+logger = logging.getLogger(__name__)
+
+
+class OlmDevice(object):
+    """Manages the Olm cryptographic functions.
+
+    Has a unique Olm account which holds identity keys.
+
+    Args:
+        api (MatrixHttpApi): The api object used to make requests.
+        user_id (str): Matrix user ID. Must match the one used when logging in.
+        device_id (str): Must match the one used when logging in.
+    """
+
+    def __init__(self, api, user_id, device_id):
+        self.api = api
+        check_user_id(user_id)
+        self.user_id = user_id
+        self.device_id = device_id
+        self.olm_account = olm.Account()
+        self.olm_account.create()
+        logger.info('Initialised Olm Device.')
+
+    def sign_json(self, json):
+        """Signs a JSON object.
+
+        As specified, this is done by encoding the JSON object without ``signatures`` or
+        keys grouped as ``unsigned``, using canonical encoding.
+
+        Args:
+            json (dict): The JSON object to sign.
+
+        Returns:
+            The same JSON object, with a ``signatures`` key added. It is formatted as
+            ``"signatures": ed25519:<device_id>: <base64_signature>``.
+        """
+        signatures = json.pop('signatures', {})
+        unsigned = json.pop('unsigned', None)
+
+        signature_base64 = self.olm_account.sign(encode_canonical_json(json))
+
+        key_id = 'ed25519:{}'.format(self.device_id)
+        signatures.setdefault(self.user_id, {})[key_id] = signature_base64.decode()
+
+        json['signatures'] = signatures
+        if unsigned:
+            json['unsigned'] = unsigned
+
+        return json
+
+    def verify_json(self, json, user_key, user_id, device_id):
+        """Verifies a signed key object's signature.
+
+        The object must have a 'signatures' key associated with an object of the form
+        `user_id: {key_id: signature}`.
+
+        Args:
+            json (dict): The JSON object to verify.
+            user_key (str): The public ed25519 key which was used to sign the object.
+            user_id (str): The user who owns the device.
+            device_id (str): The device who owns the key.
+
+        Returns:
+            True if the verification was successful, False if not.
+
+        Raises:
+            KeyError: Wrong json format.
+        """
+        signatures = json.pop('signatures')
+
+        key_id = 'ed25519:{}'.format(device_id)
+        signature_base64 = signatures[user_id][key_id]
+        unsigned = json.pop('unsigned', None)
+
+        try:
+            olm.ed25519_verify(user_key.encode(),
+                               encode_canonical_json(json),
+                               signature_base64.encode())
+            success = True
+        except olm.utility.OlmError:
```

Comment:
When running tests with olm 2.2.2 installed including python bindings from such, I get `AttributeError: module 'olm' has no attribute 'utility'`. Is this something that was added since last olm release?

Reply:
Definitly. Not only that, but also olm.ed25519_verify...

Reply:
Could you work on getting a new release of olm with the updated python bindings once you're confident they have everything needed? We won't be able to do a package release once I merge this PR until there's a stable olm this package can work with.

Of course, if we decide to use python bindings housed outside of the olm repo, it makes this a lot easier.